### PR TITLE
Refactor gateway auth

### PR DESCRIPTION
**Pull request by ableuler**
Some general cleanup of a lot of unused or non-functional code or configuration and a thorough refactoring of the flask components into 3 distinct parts:
- the web component handling the login
- the forward-auth which is called by traefik for swapping tokens
- some common code, such as classes which handle interaction with the redis store.

/deploy renku=ableuler-patch-1 #persist extra-values=gateway.traefik.logs.general.level=DEBUG

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -72,18 +66,9 @@
 
 app.config.from_object(config)
 
-CORS(
-    app,
-    allow_headers=["X-Requested-With"],
-    allow_origin=app.config["ALLOW_ORIGIN"],
-)
```
Comment (mohammad-sdsc) [inline app/__init__.py:79]: Is this not needed anymore? `ALLOW_ORIGIN` is not used anywhere else.

**Review by mohammad-sdsc — COMMENTED** [review]
Thanks Andreas! LGTM, just made a single comment.

File: app/__init__.py (hunk 1 above)
Comment (ableuler) [reply app/__init__.py:79]: indeed, all the CORS things are handled by traefik now.

Comment (ableuler) [reply app/__init__.py:79]: indeed, all the CORS things are handled by traefik now.

**Comment by ableuler** [comment]
@mohammad-sdsc This is now ready for review.

**Review by olevski — CHANGES_REQUESTED** [review]
A few changes mostly around the docs.

File: docs/diagram-draw-io.xml [hunk 2]
```diff
@@ -1 +1,2 @@
...
\ No newline at end of file
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (olevski) [reply docs/diagram-draw-io.xml:1]: btw this is the mermaid code for the image above:

```
```mermaid
sequenceDiagram
    participant User
    participant Gateway
    participant Keycloak
    participant Gitlab
    participant Redis

    User->>Gateway: /api/auth/login?redirect_url
    activate User
    activate Gateway
    Gateway-->>User: redirect
    deactivate Gateway

    User->>Keycloak: /auth/realms/Renku/protocol/openid-connect/auth
    activate Keycloak
    Keycloak-->>Keycloak: login page if needed
    Keycloak-->>User: redirect
    deactivate Keycloak
    
    User->>Gateway: /api/auth/token
    activate Gateway
    Redis->>Gateway: retrieve state, redirect_url
    Gateway->>Keycloak: /auth/realms/Renku/<br>protocol/openid-connect/token
    activate Keycloak
    Keycloak-->>Gateway: return JWT access-,<br>refresh- and id-token
    deactivate Keycloak
    Gateway->>Redis: store state, redirect url
    Gateway->>Redis: store keycloak tokens
    Gateway-->>User: redirect
    deactivate Gateway

    User->>Gateway: /api/auth/gitlab/login?redirect_url
    activate Gateway
    Gateway->>Redis: store state, redirect url
    Gateway-->>User: redirect
    deactivate Gateway
    
    User->>Gitlab: /gitlab/oauth/authorize
    activate Gitlab
    Gitlab->>Gitlab: redirect to Keycloak (KC)<br>if KC is set as identity provider<br>for Gitlab
    Gitlab-->>User: redirect
    deactivate Gitlab

    User->>Gateway: /api/auth/gitlab/token
    activate Gateway
    Redis->>Gateway: retrieve state, redirect_url
    Gateway->>Gitlab: /gitlab/oauth/token
    activate Gitlab
    Gitlab->>Gateway: return access-, refresh- and JWT id-token
    deactivate Gitlab
    Gateway->>Redis: store Gitlab tokens
    Gateway-->>User: redirect
    deactivate Gateway
    deactivate User
```

You can use [this website](https://mermaid.live/) to edit it live and see the results immediately.
```

File: app/forward_auth/forward_auth.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    """
+
+    current_app.logger.debug(f"Hitting forward-auth with args: {request.args}")
+
+    auth_case = request.args.get("auth", None)
+    if auth_case not in auth_modules:
+        current_app.logger.error(
+            f"Hitting forward-auth with missing or unknown auth case: {auth_case}"
+        )
+        return Response(f"Unknown auth case {auth_case}", status=422)
```
[the remark this answers is not in the document]
Comment (ableuler) [reply app/forward_auth/forward_auth.py]: Not really a user-provided argument, but I omitted it in the response.

File: docs/diagram-draw-io.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ableuler) [reply docs/diagram-draw-io.xml]: thanks - I've put the diagram into a separate markdown file and link to it from the main rst readme.

**Review by olevski — APPROVED** [review]

File: helm-chart/renku-gateway/values.yaml [hunk 4]
```diff
@@ -152,27 +129,6 @@ tolerations: []
 
 affinity: {}
 
-## Settings for the graph services
```
[the remark this answers is not in the document]
Comment (olevski) [reply helm-chart/renku-gateway/values.yaml:155]: I am not sure why they were removed. But I definitely cannot find any of the values that have been removed currently being used in the templates of this PR. So removing them should not break anything

**Pull request closed without merging**